### PR TITLE
Schedule builder

### PR DESCRIPTION
**Pull request by JonahZW**
### Description
Implements a mechanism to create custom schedules for every camper based on a set template. Allows for the creation, saving, and deletion of templates, as well as designating which template applies to which edah. Also provides a way to display the assignments from multiple blocks (overriding an assigned one by block, e.g. Chug Aleph for Block 1 instead of Chug Aleph for Block 2 when all other chugim shown are for Block 2).

In addition, this PR updates the entire site to use Bootstrap 5 (from Bootstrap 3).

### Specific Features
- Create a custom schedule which can be printed for each camper with their own chug assignments
- Preview the schedule before printing it
- Editor included in webpage to allow flexible formatting/editing
- Buttons enable the quick insertion of placeholders corresponding to the variables which change by camper (name, edah, chug assignments, etc)
- Overriding by block
- Save schedules for easy retrieval
    - Includes a clear way to both update a saved template and save a new template, depending on your desire
- Delete saved schedules (plus a clear warning)
- Schedules can be specifically assigned to certain edot, and be made not available for others
- Entire website overhauled to utilize Bootstrap 5 (upgraded from Bootstrap 3)

### Testing Instructions
In the directory where the repo is stored locally, run this command:
```
./start.sh
```
Then, navigate to [http://127.0.0.1:8000/](http://127.0.0.1:8000/) on that device and test normally. No coding should be required for basic testing.

### Installation/Upgrade Instructions
This update requires 2 new tables in the database:
1. Stores the schedules (id, name, schedule)
2. Stores the relation between edah and schedule to track which edot can see which schedules (schedule id, edah id)

To create these tables in the live applications of the application, sign into the SQL server, select the correct database, and run the following to create the structure which will be populated as the application runs:

```
--
-- Table structure for table `schedules`
--

DROP TABLE IF EXISTS `schedules`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!50503 SET character_set_client = utf8mb4 */;
CREATE TABLE `schedules` (
  `schedule_id` int NOT NULL AUTO_INCREMENT,
  `name` varchar(50) CHARACTER SET utf8mb3 COLLATE utf8mb3_unicode_ci NOT NULL,
  `schedule` TEXT CHARACTER SET utf8mb3 COLLATE utf8mb3_unicode_ci NOT NULL,
  PRIMARY KEY (`schedule_id`)
) ENGINE=InnoDB AUTO_INCREMENT=5 DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Table structure for table `edot_for_schedules`
--

DROP TABLE IF EXISTS `edot_for_schedule`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!50503 SET character_set_client = utf8mb4 */;
CREATE TABLE `edot_for_schedule` (
  `schedule_id` int NOT NULL,
  `edah_id` int NOT NULL,
  PRIMARY KEY (`schedule_id`,`edah_id`),
  KEY `fk_edah_id` (`edah_id`),
  CONSTRAINT `edot_for_schedule_ibfk_1` FOREIGN KEY (`schedule_id`) REFERENCES `schedules` (`schedule_id`) ON DELETE CASCADE ON UPDATE CASCADE,
  CONSTRAINT `edot_for_schedule_ibfk_2` FOREIGN KEY (`edah_id`) REFERENCES `edot` (`edah_id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;
```

### CODE REVIEW

File: chugbot/printSchedules.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    setup_camp_specific_terminology_constants();
+    echo headerText("Print Schedules");
+
+    // Get info from the form
+    $edahId = test_post_input('edah');
+    $blockId = test_post_input('block');
+    $scheduleTemplate = html_entity_decode(test_post_input('schedule-template'));
+
+    $blockIdsOverride = [];
+    for ($i = 0; $i < count($_POST)-4; $i ++) {
```
Comment (dlobron) [inline chugbot/printSchedules.php:16]: Suggest adding a comment here noting the reason for the subtraction.

File: chugbot/printSchedules.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    // ***************************************************************************
+    // **************************** Build SQL Queries ****************************
+    // ***************************************************************************
+
+    // Begin by creating a SQL query to get names of all chug groups which the
+    // selected edah has available
+    $localErr = "";
+    $dbc = new DbConn();
+    $sql = "SELECT e.group_id group_id, g.name group_name FROM edot_for_group e, chug_groups g WHERE e.edah_id IN (" . 
+        $edahId . ") AND e.group_id = g.group_id GROUP BY e.group_id HAVING COUNT(e.edah_id) = 1";
```
Comment (dlobron) [inline chugbot/printSchedules.php:36]: Here and above, I was curious about what the `HAVING` clause does. Mostly for my own knowledge.

**Comment by dlobron** [comment]
Fantastic work! I am seriously impressed by the amount of work that went into this PR!

One general question I had: could we download the Tiny editor from a CDN? Or is it large enough to make that a bad idea? I'm just thinking it's nice to get the latest version automatically.

**Comment by dlobron** [comment]
While testing locally, I noticed that the anchor links on some of the HTML pages are not formatted quite right. I've attached a screenshot example.
Screenshot 2024-05-29 at 9 13 07 PM

**Comment by dlobron** [comment]
Another minor formatting issue: on staffHome.php in the Leveling section, the names of edot and checkboxes are in a very small font:
Screenshot 2024-05-29 at 9 16 20 PM

**Comment by dlobron** [comment]
The small font issue seems to affect all checkbox groups. We might be able to fix it by just making those have a larger font in the CSS. Here is another screenshot, this one from the editChug.php. Note the small font of text in the checkboxes:

Screenshot 2024-05-29 at 9 20 58 PM:

File: chugbot/printSchedules.php (hunk 2 above)
Comment (JonahZW) [reply chugbot/printSchedules.php:36]: I think this is basically just WHERE - I believe it just verifies there is an edah... I just adapted this code from one of the reports

**Review by dlobron — APPROVED** [review]
This looks great, Jonah! Thank you for the fast turnaround.

**Pull request merged**